ldp/nn/graph/llm_call_op.py

```diff
@@ -23,71 +31,76 @@
 )
 from ..lm_config import LMConfig  # noqa: TID252
 
+logger = logging.getLogger(__name__)
 
-class LocalLLMCallOp(Op[Message]):
-    """An Op that samples a token sequence from a local language model."""
-
-    CTX_INPUTS_PREP_KEY: ClassVar[str] = "inputs_prepared"
-    CTX_TOOLS_PREP_KEY: ClassVar[str] = "tools_prepared"
-    CTX_OUTPUT_PREP_KEY: ClassVar[str] = "outputs_prepared"
 
-    model_name: str
+class MessageAndToolParser(ABC):
+    """Base class to define how we translate between (messages, tools) and strings."""
 
-    def __init__(
-        self,
-        model_config: LMConfig,
-        batch_size: int = 1,
-        max_wait_interval: float = 0.1,
-        parallel_mode_config: ParallelModeConfig | None = None,
-    ) -> None:
-        super().__init__()
+    supported_templates: ClassVar[set[str]] = set()
 
-        pad_token_id = model_config.get_tokenizer().pad_token_id
+    @classmethod
+    @abstractmethod
+    def get_message_content(cls, msg: Message) -> str | None:
+        """Represents a message as a string."""
 
-        handler_config = TransformerHandlerConfig(
-            # configurable
-            lm_config=model_config,
-            batch_size=batch_size,
-            max_wait_interval=max_wait_interval,
-            parallel_mode_config=parallel_mode_config,
-            # constant configuration
-            lm_type=LMType.GENERATION,
-            module_call_fn=AsyncTransformerInterface.model_generate,
-            collate_fn=partial(
-                collate_fn_transformer_left_pad, pad_token_id=pad_token_id
-            ),
-            decollate_fn=decollate_fn_transformer_decoder,
-        )
-        self.model_handler = handler_config.make_async_module()
-        self.model_name = model_config.model
+    @classmethod
+    @abstractmethod
+    def prep_tools_for_tokenizer(cls, tools: Tools | None) -> list[dict] | None:
```

Comment:
If we are going to prepare tools, why are we then passing `None` tools?

Just pointing out, `Tools | None` doesn't really make sense, imo this should be something like `*tools: Tool` or `tools: Tools`

Reply:
It's because `tools` can be `None` here: https://github.com/Future-House/ldp/blob/0d0e0bd27ad6e91aa271aec5c72b7ee31628a12a/ldp/graph/common_ops.py#L256 - this method just mirrors the typing of the arguments that can be passed to `[Local]LLMCallOp`.